src/main/java/org/broadinstitute/hellbender/engine/spark/SparkSharder.java

```diff
@@ -146,48 +202,68 @@ public T apply(@Nullable Tuple2<I, Iterable<L>> input) {
+                                                                                            int maxLocatableLength, FlatMapFunction2<Iterator<L>, Iterator<I>, T> f) {
+
         List<PartitionLocatable<SimpleInterval>> partitionReadExtents = computePartitionReadExtents(locatables, sequenceDictionary, maxLocatableLength);
+        List<SimpleInterval> firstLocatables = partitionReadExtents.stream().map(PartitionLocatable::getLocatable).collect(Collectors.toList());
 
         // For each interval find which partition it starts and ends in.
         // An interval is processed in the partition it starts in. However, we need to make sure that
         // subsequent partitions are coalesced if needed, so for each partition p find the latest subsequent
         // partition that is needed to read all of the intervals that start in p.
-        List<Integer> maxEndPartitionIndexes = new ArrayList<>();
-        for (int i = 0; i < locatables.getNumPartitions(); i++) {
-            maxEndPartitionIndexes.add(i);
-        }
         OverlapDetector<PartitionLocatable<SimpleInterval>> overlapDetector = OverlapDetector.create(partitionReadExtents);
-        List<PartitionLocatable<I>> indexedIntervals = new ArrayList<>();
-        int lastPartitionStartIndex = 0;
-        for (I interval : intervals) {
-            int[] partitionIndexes = overlapDetector.getOverlaps(interval).stream()
+        Broadcast<OverlapDetector<PartitionLocatable<SimpleInterval>>> overlapDetectorBroadcast = ctx.broadcast(overlapDetector);
+        JavaRDD<PartitionLocatable<I>> indexedIntervals = intervals.map(interval -> {
+            int[] partitionIndexes = overlapDetectorBroadcast.getValue().getOverlaps(interval).stream()
                     .mapToInt(PartitionLocatable::getPartitionIndex).toArray();
             if (partitionIndexes.length == 0) {
-                // interval does not overlap any partition - add it to the last one
-                indexedIntervals.add(new PartitionLocatable<>(lastPartitionStartIndex, interval));
-                continue;
+                // interval does not overlap any partition - add it to the one after the interval start
+                int i = Collections.binarySearch(firstLocatables, new SimpleInterval(interval), (o1, o2) -> IntervalUtils.compareLocatables(o1, o2, sequenceDictionary));
+                if (i >= 0) {
+                    throw new IllegalStateException(); // TODO: no overlaps, yet start of interval matches a partition read extent start
+                }
+                int insertionPoint = -i - 1;
+                if (insertionPoint == firstLocatables.size()) {
+                    insertionPoint = firstLocatables.size() - 1;
+                }
+                return new PartitionLocatable<>(insertionPoint, interval);
             }
             Arrays.sort(partitionIndexes);
             int startIndex = partitionIndexes[0];
             int endIndex = partitionIndexes[partitionIndexes.length - 1];
```

Comment:
Do we always expect it to hold that a partitioning of the reads will necessarily be in order based on the start? One could imagine pathological cases where a very tall pileup at a given start position (perhaps with some very long reads as part of it) would cause us to split the partition so there may be another partition that starts after the start of this partition but ends before the end of this partition.

Reply:
Unless this is by the start position... in which case we should better document that in computePartitionReadExetents()...

Reply:
Yes, it's by start position. Added a comment.